Added support for HTTPRoute BackendRef ports

rlymbur opened the pull request:
**What type of PR is this?**
Bug

**Which issue does this PR fix**:
#241 

**What does this PR do / Why do we need it**:
Added support for HTTPRoute BackendRef ports. This incorporates changes from both PRs, #286 and #326.

**Testing done on this change**:
Added additional E2E tests with all tests passing:
```
FOCUS="" make e2etest 
...
------------------------------

Ran 14 of 14 Specs in 5598.563 seconds
SUCCESS! -- 14 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestIntegration (5598.79s)
PASS
ok      github.com/aws/aws-application-networking-k8s/test/suites/integration   5599.360s
```

**Automation added to e2e**:
Yes, `Defined Target Ports` tests added.

**Will this PR introduce any new dependencies?**:
No

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
No, `v0.0.16` used

**Does this PR introduce any user-facing change?**:
No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mikhail-aws commented on pkg/gateway/model_build_targets.go:
```diff
@@ -111,19 +112,38 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
-		// TODO: Change the code to support multiple comma separated ports instead of a single port
-		//portsAnnotations := strings.Split(serviceExport.ObjectMeta.Annotations["multicluster.x-k8s.io/Ports"], ",")
-		portAnnotations, err = strconv.ParseInt(serviceExport.ObjectMeta.Annotations[portAnnotationsKey], 10, 64)
+	definedPorts := make([]int64, undefinedPort)
```
second argument is slice size, strange to use undefinedPort as size, you can use 0 here, or shortcut to []int64{}

zijun726911 commented on pkg/gateway/model_build_targets.go:
```diff
@@ -111,19 +112,38 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
+					}
+				}
+			}
 		}
```
You could also encapsulate line 134-143 into a helper function to make code more clear:
```
definedPorts = getDefinedPortsFromK8sRouteRules(t.route, t.tgName, t.tgNamespace)
```

```
func getDefinedPortsFromK8sRouteRules(route core.Route, tgName string, tgNamespace string) []int64 {
	if route == nil || route.Spec() == nil || route.Spec().Rules() == nil || len(route.Spec().Rules()) == 0 {
		return []int64{}
	}
	var definedPorts []int64
	for _, rule := range route.Spec().Rules() {
		for _, ref := range rule.BackendRefs() {
			if ref.Port() != nil && string(ref.Name()) == tgName && string(*ref.Namespace()) == tgNamespace {
				definedPorts = append(definedPorts, int64(*ref.Port()))
			}
		}
	}
	return definedPorts
}
``` 

(BTW,  t.tgName is actually k8sService name and t.tgNamespace is actually k8sService's namespace)

line89 `tgName := latticestore.TargetGroupName(t.tgName, t.tgNamespace)` is really confusing...

zijun726911 commented on pkg/gateway/model_build_targets.go:
```diff
@@ -111,19 +112,38 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
+				definedPorts = append(definedPorts, definedPort)
+			}
+			glog.V(6).Infof("Build Targets - portAnnotations: %v \n", definedPorts)
+		}
```
You could encapsulate line 122-132 into a helper function to make code more clear:
```
func getServiceExportAnnotationDefinedPorts(serviceExport *mcs_api.ServiceExport) []int64 {
	if serviceExport.ObjectMeta.Annotations == nil {
		return []int64{}
	}
	portsAnnotations := strings.Split(serviceExport.ObjectMeta.Annotations[portAnnotationsKey], ",")
	var definedPorts []int64
	for _, portAnnotation := range portsAnnotations {
		definedPort, err := strconv.ParseInt(portAnnotation, 10, 64)
		if err != nil {
			glog.V(6).Infof("Failed to read Annotations/Port:%v, err:%s\n ", serviceExport.ObjectMeta.Annotations[portAnnotationsKey], err)
		}
		definedPorts = append(definedPorts, definedPort)
	}
	return definedPorts
}
```

zijun726911 commented on pkg/gateway/model_build_targets.go:
```diff
@@ -111,19 +112,38 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
 		return errors.New(errmsg)
 	}
 
```
Nit: We have a giant block `if svc.DeletionTimestamp.IsZero() {}`, I think we can do a reverse condition and return earlier? e.g.,:

```
	if !svc.DeletionTimestamp.IsZero() {
		//build model model for targets deregistering
		t.latticeTargets = latticemodel.NewTargets(t.stack, tgName, latticemodel.TargetsSpec{
			Name:         t.tgName,
			Namespace:    t.tgNamespace,
			RouteName:    t.routename,
			TargetIPList: []latticemodel.Target{},
		})
	}
```

rlymbur replied:
I think the current structure has a similar effect no? Since we'll still need to build the resulting model in both cases, regardless if the timestamp is zero or now.

zijun726911 replied:
current structure has have same logic, I just think our if else loop condition too deep

rlymbur commented:
I've incorporated recent changes from `main`, on top of addressing comments, and confirmed E2E tests run again:

```
------------------------------

Ran 13 of 13 Specs in 3817.338 seconds
SUCCESS! -- 13 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestIntegration (3817.66s)
PASS
ok      github.com/aws/aws-application-networking-k8s/test/suites/integration   3818.196s
```

rlymbur commented on pkg/gateway/model_build_targets.go:
```diff
@@ -111,19 +112,38 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
+				if err != nil {
+					glog.V(6).Infof("Failed to read Annotations/Port:%v, err:%s\n ", serviceExport.ObjectMeta.Annotations[portAnnotationsKey], err)
+				}
+				definedPorts = append(definedPorts, definedPort)
```
Yes this was incorrect, I changed it now.

rlymbur commented on test/pkg/test/framework.go:
```diff
@@ -288,33 +288,43 @@ func (env *Framework) GetTargetGroup(ctx context.Context, service *v1.Service) *
+	retrievedTargets, err := env.LatticeClient.ListTargetsAsList(ctx, &vpclattice.ListTargetsInput{TargetGroupIdentifier: targetGroup.Id})
+	Expect(err).To(BeNil())
+
+	podIps := lo.Map(podList.Items, func(pod v1.Pod, _ int) string { return pod.Status.PodIP })
```
Got it, thanks for the feedback!

rlymbur commented on pkg/gateway/model_build_targets.go:
```diff
@@ -146,11 +166,18 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
+							targetList = append(targetList, target)
+							glog.V(6).Infof("portAnnotations:%v, target.Port:%v\n", definedPort, target.Port)
+						} else if t.port == undefinedPort || target.Port == t.port {
+							targetList = append(targetList, target)
```
I've revised this now, thank you!

solmonk commented on pkg/gateway/model_build_targets.go:
```diff
@@ -107,57 +108,75 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
+							// target port matches service export port
+							targetList = append(targetList, target)
+							glog.V(6).Infof("portAnnotations:%d, target.Port:%d", definedPort, target.Port)
+						} else if definedPort == undefinedPort || definedPort == target.Port {
```
I still don't get this condition, seems `definedPort == target.Port` is covered in the above condition already?

nit: the debug logs here seem a bit overkill

nit: on `// target port matches service export port` - it could be httproute not service export

solmonk commented on pkg/gateway/model_build_targets_test.go:
```diff
@@ -303,6 +313,134 @@ func Test_Targets(t *testing.T) {
 				},
 			},
 		},
+		{
```
It would be great if we could have endpoints with different ports in the test case, since that is the primary use case we want to support by this PR

solmonk commented on pkg/gateway/model_build_targets_test.go:
```diff
@@ -303,6 +313,134 @@ func Test_Targets(t *testing.T) {
+			wantErrIsNil:       false,
+		},
+		{
+			name:               "One port added via backendref match, one port added via service export annotation",
```
this might be the current behavior but we don't want to support this use case

mikhail-aws commented on pkg/gateway/model_build_targets.go:
```diff
@@ -90,13 +91,13 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
 	if err != nil {
 		errmsg := fmt.Sprintf("Build Targets failed because target group (name=%s, namespace=%s found not in datastore)", t.tgName, t.tgNamespace)
-		glog.V(6).Infof("errmsg %s\n ", errmsg)
+		glog.V(6).Infof("errmsg %s", errmsg)
```
dont log when you return error

mikhail-aws commented on pkg/gateway/model_build_targets.go:
```diff
@@ -171,14 +190,28 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
 	return nil
 }
 
+func getDefinedPortsFromK8sRouteRules(t *latticeTargetsModelBuildTask, definedPorts []int64) []int64 {
+	if t.route != nil && t.route.Spec().Rules() != nil {
+		definedPorts = []int64{}
+		for _, rule := range t.route.Spec().Rules() {
+			for _, ref := range rule.BackendRefs() {
+				if ref.Port() != nil && string(ref.Name()) == t.tgName && string(*ref.Namespace()) == t.tgNamespace {
+					definedPorts = append(definedPorts, int64(*ref.Port()))
+				}
+			}
+		}
+	}
+	return definedPorts
+}
+
```
good to see little functions with clear in-out, please add unit test for this

rlymbur replied:
This was removed now.

zijun726911 commented on test/suites/integration/httproute_header_match_test.go:
```diff
@@ -18,6 +18,11 @@ import (
+	AfterEach(func() {
+		testFramework.CleanTestEnvironment(ctx)
+	})
+
```
do you need to add this AfterEach()?

rlymbur replied:
Yes I was having resources leak on this previously.

zijun726911 commented on pkg/gateway/model_build_targetgroup.go:
```diff
@@ -145,13 +145,20 @@ func (t *latticeServiceModelBuildTask) buildTargets(ctx context.Context) error {
+			if httpBackendRef.Port() != nil {
+				port = int64(*httpBackendRef.Port())
+			}
+
```
Did you still use targetTask.port any where in current version code?

rlymbur commented:
Ran E2E tests successfully again:

```
Ran 13 of 13 Specs in 3826.191 seconds
SUCCESS! -- 13 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestIntegration (3826.39s)
PASS
ok      github.com/aws/aws-application-networking-k8s/test/suites/integration   3826.934s
```
